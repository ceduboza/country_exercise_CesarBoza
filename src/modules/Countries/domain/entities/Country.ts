import type { ICountryUI } from "./ICountryUI";

export class Country implements ICountryUI {
    private _name: string;
    private _capital: string;
    private _population: number;
    private _region: string;
    private _flag: string;

    constructor(name: string, capital: string, population: number, region: string, flag: string) {
        this._name = name;
        this._capital = capital;
        this._population = population;
        this._region = region;
        this._flag = flag;
    }

    get name(): string {
        return this._name;
    }

    get capital(): string {
        return this._capital;
    }

    get population(): number {
        return this._population;
    }

    get region(): string {
        return this._region;
    }

    get flag(): string {
        return this._flag;
    }
}