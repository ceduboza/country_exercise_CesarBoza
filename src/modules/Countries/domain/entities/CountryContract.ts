type CountryName = {
    common: string
}

export interface CountryContract {
    name: CountryName,
    capital: string[]
    flag: string
    population: number
    region: string
}