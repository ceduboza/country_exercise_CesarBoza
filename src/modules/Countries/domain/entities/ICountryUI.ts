export interface ICountryUI {
    name: string;
    capital: string;
    population: number;
    region: string;
    flag: string;
}
