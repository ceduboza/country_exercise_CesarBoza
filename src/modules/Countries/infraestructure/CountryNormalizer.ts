import type { Country } from "src/modules/Countries/domain/entities/Country";
import type { CountryContract } from "src/modules/Countries/domain/entities/CountryContract";

export class CountryNormalizer {
    normalize(country: CountryContract[]): Country[] {
        return country.map((country: CountryContract) => ({
            name: country.name?.common || '',
            capital: country.capital?.[0] || '',
            flag: country?.flag || '',
            population: country.population || 0,
            region: country.region || '',
        }))
    }
}