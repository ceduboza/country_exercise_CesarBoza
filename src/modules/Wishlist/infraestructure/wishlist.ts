import { useStorage } from '@vueuse/core'
import type { ICountryUI } from 'src/modules/Countries/domain/entities/ICountryUI'

export const useWishlist = () => {
    const wishlist = useStorage('wishlist', [] as ICountryUI[])

    const addToWishlist = (country: ICountryUI) => {
        wishlist.value.push({
            name: country.name,
            capital: country.capital,
            population: country.population,
            region: country.region,
            flag: country.flag
        })
    }

    const removeFromWishlist = (country: ICountryUI) => {
        wishlist.value = wishlist.value.filter((wishlistCountry: ICountryUI) => wishlistCountry.name !== country.name)
    }

    const isCountryInWishlist = (country: ICountryUI) => {
        return wishlist.value.some((wishlistCountry: ICountryUI) => wishlistCountry.name === country.name)
    }

    return {
        wishlist,
        addToWishlist,
        removeFromWishlist,
        isCountryInWishlist
    }
}
