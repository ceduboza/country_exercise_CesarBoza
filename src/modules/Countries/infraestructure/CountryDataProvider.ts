import type { Country } from "../domain/entities/Country";
import type { CountryContract } from "../domain/entities/CountryContract";
import type { ICountryRepository } from "../domain/repositories/ICountryRepository";
import { CountryNormalizer } from "./CountryNormalizer";

export class CountryDataProvider implements ICountryRepository {
    private _normalizer: CountryNormalizer;

    constructor() {
        this._normalizer = new CountryNormalizer();
    }

    getCountries(): Promise<Country[]> {
        const params = new URLSearchParams({
            fields: 'name,capital,population,region,flag'
        }).toString();

        return fetch(`https://restcountries.com/v3.1/all?${params}`, { method: 'GET' })
            .then(response => response.json())
            .then(data => this._normalizer.normalize(data as CountryContract[]));
    }

    getCountriesByName(name: string): Promise<Country[]> {
        return fetch(`https://restcountries.com/v3.1/name/${name}`, { method: 'GET' })
            .then(response => response.json())
            .then(data => this._normalizer.normalize(data as CountryContract[]));
    }
}
