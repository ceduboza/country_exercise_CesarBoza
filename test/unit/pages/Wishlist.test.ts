import { describe, it, expect, vi } from "vitest";
import { shallowMount } from "@vue/test-utils";
import WishlistPage from "../../../src/pages/Wishlist/index.vue";
import { useWishlist } from "../../../src/modules/Wishlist/infraestructure/wishlist";
import { ICountryUI } from "../../../src/modules/Countries/domain/entities/ICountryUI";

vi.mock("../../../src/modules/Wishlist/infraestructure/wishlist", () => ({
    useWishlist: vi.fn()
}))

interface ContextMock {
    wishlist: ICountryUI[]
}

describe("Wishlist", () => {
    let wrapper

    const findQpage = wrapper => wrapper.findComponent({ name: 'Qpage' })
    const findHeaderPage = wrapper => wrapper.findComponent({ name: 'HeaderPage' })
    const findCountryGrid = wrapper => wrapper.findComponent({ name: 'CountryGrid' })
    const findEmptyState = wrapper => wrapper.findComponent({ name: 'EmptyState' })
    const findWishlistIcon = wrapper => wrapper.findComponent({ name: 'WishlistIcon' })

    const contextFactory = (contextMock?: ContextMock) => {
        vi.mocked(useWishlist).mockReturnValue({
            wishlist: contextMock?.wishlist || []
        } as any)
    }

    const factory = (contextMock?: ContextMock) => {
        contextFactory(contextMock)

        wrapper = shallowMount(WishlistPage, {
            global: {
                stubs: {
                    QPage: {
                        name: 'Qpage',
                        template: '<div unit-test-id="q-page"><slot /></div>',
                        props: ['title']
                    },
                    CountryGrid: {
                        name: 'CountryGrid',
                        template: `<div unit-test-id="country-grid"><slot name="actions" :country='{ "name": "Test Country" }' /></div>`,
                    },
                    HeaderPage: true,
                    EmptyState: true,
                    WishlistIcon: true
                }
            }
        })
    }

    it("should render the wishlist page if wishlist is empty", () => {
        factory()

        expect(wrapper.html()).toMatchSnapshot()
        expect(findQpage(wrapper).exists()).toBeTruthy()
        expect(findEmptyState(wrapper).exists()).toBeTruthy()
        expect(findCountryGrid(wrapper).exists()).toBeFalsy()
        expect(findHeaderPage(wrapper).exists()).toBeTruthy()
        expect(findWishlistIcon(wrapper).exists()).toBeFalsy()
    })
    it("should render the wishlist page if wishlist is not empty", () => {
        const contextMock = {
            wishlist: [{
                name: "Country",
                capital: "Capital",
                population: 1000000,
                region: "Region",
                flag: "Flag"
            }]
        }
        factory(contextMock)

        expect(wrapper.html()).toMatchSnapshot()
        expect(findQpage(wrapper).exists()).toBeTruthy()
        expect(findEmptyState(wrapper).exists()).toBeFalsy()
        expect(findCountryGrid(wrapper).exists()).toBeTruthy()
        expect(findHeaderPage(wrapper).exists()).toBeTruthy()
        expect(findWishlistIcon(wrapper).exists()).toBeTruthy()
    })
})
