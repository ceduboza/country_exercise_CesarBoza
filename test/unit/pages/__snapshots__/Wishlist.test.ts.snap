// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wishlist > should render the wishlist page if wishlist is empty 1`] = `
"<div unit-test-id="q-page" class="wishlist-page">
  <header-page-stub title="Wishlist"></header-page-stub>
  <empty-state-stub title="No countries in wishlist" icon="favorite" class="wishlist-page__empty-state"></empty-state-stub>
</div>"
`;

exports[`Wishlist > should render the wishlist page if wishlist is not empty 1`] = `
"<div unit-test-id="q-page" class="wishlist-page">
  <header-page-stub title="Wishlist"></header-page-stub>
  <div unit-test-id="country-grid" countries="[object Object]" isloading="false">
    <wishlist-icon-stub country="[object Object]"></wishlist-icon-stub>
  </div>
</div>"
`;
